pkg/coscheduling/core/core.go

```diff
@@ -0,0 +1,241 @@
+	pginformer "sigs.k8s.io/scheduler-plugins/pkg/generated/informers/externalversions/scheduling/v1alpha1"
+	pglister "sigs.k8s.io/scheduler-plugins/pkg/generated/listers/scheduling/v1alpha1"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+)
+
+type Manager interface {
+	PreFilter(context.Context, *corev1.Pod) error
+	Permit(context.Context, *corev1.Pod, string) (bool, error)
+	PostBind(context.Context, *corev1.Pod, string)
+	GetPodGroup(*corev1.Pod) (string, *v1alpha1.PodGroup)
+	GetCreationTime(*corev1.Pod, time.Time) time.Time
+	AddToDenyCache(string)
+}
+
+// PodGroupManager defines the scheduling operation called
+type PodGroupManager struct {
+	// pgClient is a podGroup client
+	pgClient pgclientset.Interface
+	// snapshotSharedLister is pod shared list
+	snapshotSharedLister framework.SharedLister
+	// scheduleTimeout is the default time when group scheduling.
+	// If podgroup's ScheduleTimeoutSeconds set, that would be used.
+	scheduleTimeout *time.Duration
+	// lastDeniedPG store the pg name if a pod can not pass pre-filer,
+	// or anyone of the pod timeout
+	lastDeniedPG *gochache.Cache
+	// pgLister is podgroup lister
+	pgLister pglister.PodGroupLister
+	// podLister is pod lister
+	podLister listerv1.PodLister
+	// reserveResourcePercentage is the reserved resource for the max finished group, range (0,100]
+	reserveResourcePercentage int32
+	sync.RWMutex
+}
+
+// NewPodGroupManager create a new operation object
+func NewPodGroupManager(pgClient pgclientset.Interface, snapshotSharedLister framework.SharedLister, scheduleTimeout *time.Duration,
+	pgInformer pginformer.PodGroupInformer, podInformer informerv1.PodInformer) *PodGroupManager {
+	pgMgr := &PodGroupManager{
+		pgClient:             pgClient,
+		snapshotSharedLister: snapshotSharedLister,
+		scheduleTimeout:      scheduleTimeout,
+		pgLister:             pgInformer.Lister(),
+		podLister:            podInformer.Lister(),
+		lastDeniedPG:         gochache.New(3*time.Second, 3*time.Second),
+	}
+	return pgMgr
+}
+
+// PreFilter pre-filter pod include:
+// 1. cluster resource
+// 2. last denied pod to increase schedule qps
+// 3. check resource for max finished group, evict others if it can not run
+// 4. check pod priority, if it can not run, evict other groups
+func (pgMgr *PodGroupManager) PreFilter(ctx context.Context, pod *corev1.Pod) error {
+	klog.V(10).Infof("Pre-filter %v", pod.Name)
+	pgFullName, pg := pgMgr.GetPodGroup(pod)
+	if pg == nil {
+		return nil
+	}
+	if _, ok := pgMgr.lastDeniedPG.Get(pgFullName); ok {
+		err := fmt.Errorf("pod with pgName: %v last failed in 20s, deny", pgFullName)
+		klog.V(6).Info(err)
+		return err
+	}
+	pods, err := pgMgr.podLister.Pods(pod.Namespace).List(
+		labels.SelectorFromSet(labels.Set{util.PodGroupLabel: util.GetPodGroupLabel(pod)}),
+	)
+	if err != nil {
+		return fmt.Errorf("podLister list pods failed: %v", err)
+	}
+	if len(pods) < int(pg.Spec.MinMember) {
```

Comment:
I suggest to remove this judgment. This logic is too easy to trigger.

Reply:
Here we just keep the original logic.

IMO, this may happen frequently during creating pods. But if we remove this, we may schedule a pod and add to waitingPod(which would occupy the resource)  though number of pods can not meet the min Member.

Reply:
Yes, it's the current behavior:

https://github.com/kubernetes-sigs/scheduler-plugins/blob/8dec3f44981d9252c51590577c72f45597b01001/pkg/coscheduling/coscheduling.go#L229